waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/ParallelActionScheduler.scala

```diff
@@ -0,0 +1,106 @@
+package com.coxautodata.waimak.dataflow
+
+import java.util
+import java.util.concurrent.{BlockingQueue, Executors, LinkedBlockingQueue, TimeUnit}
+
+import com.coxautodata.waimak.log.Logging
+
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
+import scala.util.{Failure, Success, Try}
+import scala.collection.JavaConverters._
+
+/**
+  * Created by Alexei Perelighin on 2018/07/10
+  *
+  * @param pools
+  * @tparam C
+  */
+class ParallelActionScheduler[C](val pools: Map[String, ExecutionPoolDesc]
+                                 , val actionFinishedNotificationQueue: BlockingQueue[(String, DataFlowAction[C], Try[ActionResult])]
+                                , val poolIntoContext: (String, C) => Unit
+                                )
+  extends ActionScheduler[C] with Logging {
+
+  type actionType = DataFlowAction[C]
+
+  type futureResult = (String, actionType, Try[ActionResult])
+
+  override def dropRunning(poolNames: Set[String], from: Seq[DataFlowAction[C]]): Seq[DataFlowAction[C]] = {
+    if (from.isEmpty) from
+    else {
+      val running = pools.filterKeys(poolNames.contains).values.flatMap(_.running).toSet
+      from.filter(a => !running.contains(a.schedulingGuid))
+    }
+  }
+
+  override def hasRunningActions: Boolean = pools.exists(kv => kv._2.running.nonEmpty)
+
+  override def waitToFinish(): Try[(ActionScheduler[C], Seq[(DataFlowAction[C], Try[ActionResult])])] = {
+    Try {
+      var finished: Option[Seq[futureResult]] = None
+      do {
+        val runningActionsCount = pools.map(_._2.running.size).sum
+        logInfo(s"Waiting for an action to finish to continue. Running actions: $runningActionsCount")
+        finished = Option(actionFinishedNotificationQueue.poll(1, TimeUnit.SECONDS))
+          .map { oneAction =>
+            //optimistic attempt to get results for more actions without blocking
+            val bucket = new util.LinkedList[futureResult]()
+            actionFinishedNotificationQueue.drainTo(bucket, 1000)
+            bucket.asScala :+ oneAction
+          }
+      } while (finished.isEmpty)
+      logInfo(s"No more waiting for an action. ${finished.isDefined}")
+      finished.map { rSet =>
+        val poolActionGuids: Map[String, Set[String]] = rSet.map(r => (r._1, r._2.schedulingGuid)).groupBy(_._1).mapValues(v => v.map(_._2).toSet)
+        val newPools = poolActionGuids.foldLeft(pools) { (newPools, kv) =>
+          val newPoolDesc = newPools(kv._1).removeActionGUIDS(kv._2)
+          newPools + (kv._1 -> newPoolDesc)
+        }
+        (new ParallelActionScheduler(newPools, actionFinishedNotificationQueue, poolIntoContext), rSet.map(r => (r._2, r._3)))
+      }
+    }.flatMap {
+      case Some(r) => Success(r)
+      case _ => Failure(new DataFlowException(s"Something when wrong!!! Not sure what happend."))
```

Comment:
Spelling, also can we give more context in the error?

Reply:
done